validator-api/src/epoch_operations/mod.rs

```diff
@@ -88,47 +78,79 @@ impl RewardedSetUpdater {
         })
     }
 
+    // Needs to run for active and reserve sets separatley, as it does not preserve order
     async fn determine_layers(
         &self,
-        rewarded_set: &[MixNodeDetails],
-    ) -> Result<(Vec<LayerAssignment>, HashMap<String, Layer>), RewardingError> {
-        let mut families_in_layer: HashMap<String, Layer> = HashMap::new();
-        let mut assignments = vec![];
-        let mut layer_assignments: HashMap<Layer, f32> = HashMap::new();
-        let mut rng = OsRng;
-        let layers = vec![Layer::One, Layer::Two, Layer::Three];
+        set: &[MixNodeDetails],
+    ) -> Result<Vec<LayerAssignment>, RewardingError> {
+        let mut assignments = Vec::with_capacity(set.len());
+        let target_layer_count = set.len() / 3;
 
         let mix_to_family = self.validator_cache.mix_to_family().await.to_vec();
 
         let mix_to_family = mix_to_family
             .into_iter()
             .collect::<HashMap<IdentityKey, FamilyHead>>();
 
-        for mix in rewarded_set {
-            let family = mix_to_family.get(&mix.bond_information.identity().to_owned());
-            // Get layer already assigned to nodes family, if any
-            let family_layer = family.and_then(|h| families_in_layer.get(h.identity()));
+        let mut regular_nodes = Vec::with_capacity(set.len());
+
+        let mut families = HashMap::new();
 
-            // Same node families are always assigned to the same layer, otherwise layer selected by a random weighted choice
-            let layer = if let Some(layer) = family_layer {
-                layer.to_owned()
+        for node in set.iter() {
+            if let Some(fh) = mix_to_family.get(node.bond_information.identity()) {
+                let family: &mut Vec<u32> = families.entry(fh.identity()).or_default();
+                family.push(node.mix_id())
             } else {
-                layers
-                    .choose_weighted(&mut rng, |l| layer_weight(l, &layer_assignments))?
-                    .to_owned()
-            };
+                regular_nodes.push(node.mix_id())
+            }
+        }
 
-            assignments.push(LayerAssignment::new(mix.mix_id(), layer));
+        let mut layers = HashMap::new();
+        layers.insert(Layer::One, Vec::with_capacity(target_layer_count));
+        layers.insert(Layer::Two, Vec::with_capacity(target_layer_count));
+        layers.insert(Layer::Three, Vec::with_capacity(target_layer_count));
+
+        // Assign all members of a family to same layer
+        for (_head, members) in families.iter_mut() {
+            let smallest_layer = layers
+                .iter()
+                .min_by_key(|(_layer, members)| members.len())
+                .map(|(layer, _members)| *layer)
+                .unwrap_or(Layer::One);
+
+            // What if a single family is bigger then the layer
+            if members.len() > target_layer_count {
+                members.truncate(target_layer_count)
```

Comment:
but isn't it the case that after we truncate this vec, we will "lose" those members?